sdk/src/azure/platform/az_mosquitto.c

```diff
@@ -0,0 +1,278 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+/**
+ * @file
+ *
+ * @brief Contains the az_mqtt.h interface implementation with Mosquitto MQTT
+ * (https://github.com/eclipse/mosquitto).
+ *
+ * @note The Mosquitto Lib documentation is available at:
+ * https://mosquitto.org/api/files/mosquitto-h.html
+ */
+
+#include <azure/core/az_mqtt.h>
+#include <azure/core/az_platform.h>
+#include <azure/core/az_span.h>
+#include <azure/core/internal/az_log_internal.h>
+#include <azure/core/internal/az_precondition_internal.h>
+#include <azure/core/internal/az_result_internal.h>
+#include <azure/core/internal/az_span_internal.h>
+
+#include <mosquitto.h>
+#include <pthread.h>
+#include <stdlib.h>
+
+#include <azure/core/_az_cfg.h>
+
+static void _az_mosquitto_critical_error() { az_platform_critical_error(); }
+
+AZ_INLINE az_result _az_result_from_mosq(int mosquitto_ret)
+{
+  az_result ret;
+
+  if (mosquitto_ret == MOSQ_ERR_SUCCESS)
+  {
+    ret = AZ_OK;
+  }
+  else
+  {
+    ret = _az_RESULT_MAKE_ERROR(_az_FACILITY_IOT_MQTT, mosquitto_ret);
+  }
+
+  return ret;
+}
+
+static void _az_mosquitto_on_connect(struct mosquitto* mosq, void* obj, int reason_code)
+{
+  az_result ret;
+  az_mqtt* me = (az_mqtt*)obj;
+
+  _az_PRECONDITION(mosq == me->mosquitto_handle);
```

Comment:
Sync'd with Chris on the obj being `me`, and we discussed that it should be fine, but all of the callbacks should have this _az_PRECONDITION check and then not use `mosq` past creating the `me` var for consistency (ex. the mosquitto_disconnect should pass in me->mosquitto_handle instead of mosq)

Reply:
Added precondition checks + removed usage of mosq